File: 240702/마을 구분하기/seperate-village.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
#define MAX_N 25
#define DIR_NUM 4

int n, cnt;
int grid[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N];
vector<int> people;

int dx[DIR_NUM] = {-1, 1, 0, 0}, dy[DIR_NUM] = {0, 0, -1, 1};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

bool CanGo(int x, int y) {
    if(!InRange(x, y)) return false;
    if(visited[x][y]) return false;
    if(grid[x][y] == 0) return false;

    return true;
}

void DFS(int x, int y) {
    for(int i=0; i<DIR_NUM; i++) {
        int nx = x+dx[i], ny = y+dy[i];

        if(CanGo(nx, ny)) {
            visited[nx][ny] = 1;
            cnt++;
            DFS(nx, ny);
        }
    }
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
        }
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(grid[i][j] == 1 && visited[i][j] == 0) {
                visited[i][j] = 1;
                cnt = 1;
                DFS(i, j);

                people.push_back(cnt);
            }
        }
    }
    
    sort(people.begin(), people.end());

    cout << (int) people.size() << '\n';
    for(int i=0; i<people.size(); i++) {
        cout << people[i] << '\n';
    }

    return 0;
}

File: 250926/마을 구분하기/seperate-village.py
n = int(input())

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

visited = [[False]*n for _ in range(n)]
village = []

cnt = 0

def can_go(x, y):
    return 0<=x<n and 0<=y<n and grid[x][y] and not visited[x][y]

def dfs(x, y):
    global cnt

    for dx, dy in zip(dxs, dys):
        nx, ny = x+dx, y+dy

        if can_go(nx, ny):
            visited[nx][ny] = True
            cnt += 1
            dfs(nx, ny)

for i in range(n):
    for j in range(n):
        if not visited[i][j] and grid[i][j]:
            visited[i][j] = True
            cnt = 1
            dfs(i, j)

            village.append(cnt)

village.sort()

print(len(village))
for v in village:
    print(v)